src/ripgrep-directory-searcher.js

```diff
@@ -0,0 +1,187 @@
+const { spawn } = require('child_process')
+
+// `ripgrep` and `scandal` have a different way of handling the trailing and leading
+// context lines:
+//  * `scandal` returns all the context lines that are requested, even if they include
+//    previous or future results.
+//  * `ripgrep` is a bit smarter and only returns the context lines that do not correspond
+//    to any result (in a similar way that is shown in the find and replace UI).
+//
+// For example, if we have the following file and we request to leading context lines:
+//
+//    line 1
+//    line 2
+//    result 1
+//    result 2
+//    line 3
+//    line 4
+//
+// `scandal` will return two results:
+//   * First result with `['line 1', line 2']` as leading context.
+//   * Second result with `['line 2', result 1']` as leading context.
+// `ripgrep` on the other hand will return a JS object that is more similar to the way that
+// the results are shown:
+//   [
+//     {type: 'begin', ...},
+//     {type: 'context', ...}, // context for line 1
+//     {type: 'context', ...}, // context for line 2
+//     {type: 'match', ...}, // result 1
+//     {type: 'match', ...}, // result 2
+//     {type: 'end', ...},
+//   ]
+//
+// In order to keep backwards compatibility, and avoid doing changes to the find and replace logic,
+// for `ripgrep` we need to keep some state with the context lines (and matches) to be able to build
+// a data structure that has the same behaviour as the `scandal` one.
+//
+// We use the `pendingLeadingContext` array to generate the leading context. This array gets mutated
+// to always contain the leading `n` lines and is cloned every time a match is found. It's currently
+// implemented as a standard array but we can easily change it to use a linked list if we find that
+// the shift operations are slow.
+//
+// We use the `pendingTrailingContexts` Set to generate the trailing context. Since the trailing
+// context needs to be generated after receiving a match, we keep all trailing context arrays that
+// haven't been fulfilled in this Set, and mutate them adding new lines until they are fulfilled.
+
+function updateLeadingContext (message, pendingLeadingContext, options) {
+  if (message.type !== 'match' && message.type !== 'context') {
+    return
+  }
+
+  if (options.leadingContextLineCount) {
+    pendingLeadingContext.push(message.data.lines.text.trim())
+
+    if (pendingLeadingContext.length > options.leadingContextLineCount) {
+      pendingLeadingContext.shift()
+    }
+  }
+}
+
+function updateTrailingContexts (message, pendingTrailingContexts, options) {
+  if (message.type !== 'match' && message.type !== 'context') {
+    return
+  }
+
+  if (options.trailingContextLineCount) {
+    for (const trailingContextLines of pendingTrailingContexts) {
+      trailingContextLines.push(message.data.lines.text.trim())
```

Comment:
What if `message.data.lines.text` contains newlines?

Reply:
I haven't tested this, but I'm going to add a test for that.

I remember that when we were testing the current implementation we saw some issues with multi-line results right?

Reply:
Yeah, although I honestly can't remember what we saw.